### ISSUE TITLE
Have module to opt-in Hamlit::Block::Engine

### ISSUE THREAD

**Issue by k0kubun** [trailblazer/cells-hamlit#6]
I changed hamlit-block not to alter `Hamlit::Engine`'s behavior and to have `Hamlit::Block::Engine` with hamlit-block's original behavior independently.

We should create a module to opt-in `Hamlit::Block::Engine`'s feature like: https://github.com/trailblazer/cells-hamlit/pull/2#issuecomment-185169764

The usage of `Hamlit::Block::Engine` is the same as `Hamlit::Engine`. All you need is `Hamlit::Block::Engine.new.call(template)`.

```rb
irb> puts Hamlit::Block::Engine.new.call("hello\n- this = is do\n  a beaultful\n  world!\n")
_buf = []; _buf << ("hello\n".freeze);
;  this = is do;
; _hamlit_compiler1 = ''; _hamlit_compiler1 << ("a beaultful\nworld!\n".freeze);
; _hamlit_compiler1; end; _buf = _buf.join("".freeze)
```

I shipped hamlit-block v0.4.0 with this feature. But it won't work with cells-hamlit <= v0.2.0. We should create the module and ship cells-hamlit with v0.3.0.

**Comment by k0kubun** [comment]
NOTE: We can add hamlit-block v0.4.0 as dependency, not optional. For that, I intentionally avoided adding cells-hamlit as dependency of hamlit-block not to create circular dependency https://github.com/hamlit/hamlit-block/commit/92b15fbf7cc65e34b617a4f6d2584123300b8ad2. Please start hacking with incremented `Cell::Hamlit::VERSION`.

**Comment by k0kubun** [comment]
@apotonick Hi, could you give me some advice to do this?

**Comment by apotonick** [comment]
Cool, so should cells-hamlit use the `Block::Engine`?

**Comment by k0kubun** [comment]
> Cool, so should cells-hamlit use the Block::Engine?

Yes

**Comment by apotonick** [comment]
With `Block`, `for` loops etc. don't work anymore, right?

**Comment by k0kubun** [comment]
> for loops etc. don't work anymore

Any loop should call yield to render the contents. In that meaning, yes. But it should be an original behavior  of hamlit-block.

**Comment by apotonick** [comment]
Now I'm confused, can you show an example how to use a loop with the block semantics?

**Comment by k0kubun** [comment]
All rendering results of nested block are captured and returned from the nested block. It's the only way to achieve capturing with yield. So every loop's block result should be returned by `map` and you should render them from a unnested script.

The following template:

```haml
%ul
  - htmls = [1,2,3].map do |i|
    %li
      = i
  = htmls.join("\n")
```

should be rendered to:

```html
<ul>
<li>
1
</li>
<li>
2
</li>
<li>
3
</li>
</ul>
```

Capturing by `yield` means that you should treat all block's result by yourself.

**Comment by apotonick** [comment]
Ok I understand that, but Cells-hamlit users then would have to use that syntax from the template you provided?

**Comment by k0kubun** [comment]
Yes. But I think Slim users should do this if Slim is also capturable by `yield`. If it is not the truth, please tell me what is expected code in Slim to be used. I can research how it is achieved and it may be easy to reproduce Slim's behavior.

For example, `for` loop's nested scripts must not be blocks. And we can write some implementation to distinguish block and `for` loop. So I can change the behavior inside `for` loop to be renderable.

**Comment by k0kubun** [comment]
Regardless of that problem, we should resolve this issue separately because cells-hamlit v0.1.1 and v0.2.0 are the same behavior with `Hamlit::Block::Engine` and this problem has existed since cells-hamlit v0.1.1 is released. 

And recent hamlit-block doesn't modify Hamlit's behavior globally. It means that currently only cells-hamlit v0.1.0 is usable with the latest hamlit-block.

**Comment by k0kubun** [comment]
Is there any example that replaces rendering engine by including cells module?

**Comment by PikachuEXE** [comment]
@k0kubun 
Can you take a look on #7? Thanks

**Comment by emptyflask** [comment]
Any progress on this in the past 5 years?
Should cells-rails continue to be used, or would it be better to abandon it for `view_component`?

**Comment by k0kubun** [comment]
As PiachuEXE pointed out in their PR, you could use cells-hamlit2 for now. If it doesn't work, you may need to find something else.

**Comment by emptyflask** [comment]
https://github.com/hamlit/hamlit-block/blob/master/lib/hamlit/block/version.rb#L9-L10

This is the exception i get when trying to use hamlit-block, even with cells-hamlit2... Perhaps I should fork it and see what happens when I bypass this check

**Comment by k0kubun** [comment]
You don't need hamlit-block if you use cells-hamlit2. Just uninstall it.

**Comment by apotonick** [comment]
@k0kubun Hey!!! :heart_eyes: Remind me, why is it that we maintain two versions of this gem, again?

**Comment by k0kubun** [comment]
Honestly, I can't remember and explain the exact details at this point :slightly_smiling_face: I believe it's something like this: cells-hamlit tries to implement block rendering with `yield` whereas cells-hamlit2 follows what Rails does, and `yield` can't handle some cases well.

If we don't like having two versions, I guess we could change cells-hamlit to use the implementation of cells-hamlit2.

**Comment by apotonick** [comment]
Hamlit uses `output_buffer`, right? :no_good:

**Comment by k0kubun** [comment]
It uses an array buffer by default, but it uses OutputBuffer for Rails, yes.

**Comment by k0kubun** [comment]
Today I released Hamlit v3.0.0, which works like hamlit-block by default for non-Rails environments. Given that cells-hamlit uses `Hamlit::Template` (non-Rails one), I think it's now enabled for cells-hamlit by default too. Please try it when you have a chance.
